cirq-google/cirq_google/engine/abstract_job.py

```diff
@@ -0,0 +1,201 @@
+    execution of a calibration request.
+
+    This job may be in a variety of states. It may be scheduling, it may be
+    executing on a machine, or it may have entered a terminal state
+    (either succeeding or failing).
+
+    `AbstractJob`s can be iterated over, returning `Result`s. These
+    `Result`s can also be accessed by index. Note that this will block
+    until the results are returned.
+
+    """
+
+    @abstractmethod
+    def engine(self) -> 'AbstractEngine':
+        """Returns the parent `AbstractEngine` object."""
+
+    @abstractmethod
+    def id(self) -> str:
+        """Returns the id of this job."""
+
+    @abstractmethod
+    def program(self) -> 'AbstractProgram':
+        """Returns the parent `AbstractProgram`object."""
+
+    @abstractmethod
+    def create_time(self) -> 'datetime.datetime':
+        """Returns when the job was created."""
+
+    @abstractmethod
+    def update_time(self) -> 'datetime.datetime':
+        """Returns when the job was last updated."""
+
+    @abstractmethod
+    def description(self) -> str:
+        """Returns the description of the job."""
+
+    @abstractmethod
+    def set_description(self, description: str) -> 'AbstractJob':
+        """Sets the description of the job.
+
+        Params:
+            description: The new description for the job.
+
+        Returns:
+             This `AbstractJob`.
+        """
+
+    @abstractmethod
+    def labels(self) -> Dict[str, str]:
+        """Returns the labels of the job."""
+
+    @abstractmethod
+    def set_labels(self, labels: Dict[str, str]) -> 'AbstractJob':
+        """Sets (overwriting) the labels for a previously created quantum job.
+
+        Params:
+            labels: The entire set of new job labels.
+
+        Returns:
+             This `AbstractJob`.
+        """
+
+    @abstractmethod
+    def add_labels(self, labels: Dict[str, str]) -> 'AbstractJob':
+        """Adds new labels to a previously created quantum job.
+
+        Params:
+            labels: New labels to add to the existing job labels.
+
+        Returns:
+             This `AbstractJob`.
+        """
+
+    @abstractmethod
+    def remove_labels(self, keys: List[str]) -> 'AbstractJob':
+        """Removes labels with given keys.
+
+        Params:
+            label_keys: Label keys to remove from the existing job labels.
+
+        Returns:
+            This `AbstractJob`.
+        """
+
+    @abstractmethod
+    def processor_ids(self) -> List[str]:
+        """Returns the processor ids provided when the job was created."""
+
+    @abstractmethod
+    def execution_status(self) -> quantum.enums.ExecutionStatus.State:
+        """Return the execution status of the job."""
+
+    @abstractmethod
+    def failure(self) -> Optional[Tuple[str, str]]:
+        """Return failure code and message of the job if present."""
+
+    @abstractmethod
+    def get_repetitions_and_sweeps(self) -> Tuple[int, List[cirq.Sweep]]:
```

Comment:
Should `List[cirq.Sweep]` instead be `cirq.Sweepable`? (I'm not certain on this - there's a bunch of types in the `Sweep` space, and I'm not 100% clear on their hierarchy)

Reply:
SGTM.